File: web/src/api/endpoints/route.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '../client';
import { logger } from '@/lib/logger';

/**
 * 工作分组信息
 */
export interface RouteGroup {
    id?: number;
    route_id?: number;
    group_id: number;
    category: string;
    description?: string;
    keywords?: string;
}

/**
 * 路由信息
 */
export interface Route {
    id?: number;
    name: string;
    primary_group_id: number;
    dispatch_group_id?: number;
    description?: string;
    work_groups?: RouteGroup[];
}

/**
 * 新增工作分组请求
 */
export interface RouteGroupAddRequest {
    group_id: number;
    category: string;
    description?: string;
    keywords?: string;
}

/**
 * 更新工作分组请求
 */
export interface RouteGroupUpdateRequest {
    id: number;
    group_id?: number;
    category?: string;
    description?: string;
    keywords?: string;
}

/**
 * 创建路由请求
 */
export interface RouteCreateRequest {
    name: string;
    primary_group_id: number;
    dispatch_group_id?: number;
    description?: string;
    work_groups?: RouteGroupAddRequest[];
}

/**
 * 更新路由请求 - 仅包含变更的数据
 */
export interface RouteUpdateRequest {
    id: number;
    name?: string;
    primary_group_id?: number;
    dispatch_group_id?: number;
    description?: string;
    work_groups_to_add?: RouteGroupAddRequest[];
    work_groups_to_update?: RouteGroupUpdateRequest[];
    work_groups_to_delete?: number[];
}

/**
 * 获取路由列表 Hook
 */
export function useRouteList() {
    return useQuery({
        queryKey: ['routes', 'list'],
        queryFn: async () => {
            return apiClient.get<Route[]>('/api/v1/route/list');
        },
        refetchInterval: 30000,
        refetchOnMount: 'always',
    });
}

/**
 * 创建路由 Hook
 */
export function useCreateRoute() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (data: RouteCreateRequest) => {
            return apiClient.post<Route>('/api/v1/route/create', data);
        },
        onSuccess: (data) => {
            logger.log('路由创建成功:', data);
            queryClient.invalidateQueries({ queryKey: ['routes', 'list'] });
        },
        onError: (error) => {
            logger.error('路由创建失败:', error);
        },
    });
}

/**
 * 更新路由 Hook - 仅发送变更的数据
 */
export function useUpdateRoute() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (data: RouteUpdateRequest) => {
            return apiClient.post<Route>('/api/v1/route/update', data);
        },
        onSuccess: (data) => {
            logger.log('路由更新成功:', data);
            queryClient.invalidateQueries({ queryKey: ['routes', 'list'] });
        },
        onError: (error) => {
            logger.error('路由更新失败:', error);
        },
    });
}

/**
 * 删除路由 Hook
 */
export function useDeleteRoute() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (id: number) => {
            return apiClient.delete<null>(`/api/v1/route/delete/${id}`);
        },
        onSuccess: () => {
            logger.log('路由删除成功');
            queryClient.invalidateQueries({ queryKey: ['routes', 'list'] });
        },
        onError: (error) => {
            logger.error('路由删除失败:', error);
        },
    });
}

File: internal/model/route.go
package model

// Route 路由 - 分组之上的独立层，根据请求内容自动选择分组
type Route struct {
	ID               int          `json:"id" gorm:"primaryKey"`
	Name             string       `json:"name" gorm:"unique;not null"`
	PrimaryGroupID   int          `json:"primary_group_id" gorm:"not null"`      // 主分组 ID（必选，兜底）
	DispatchGroupID  *int         `json:"dispatch_group_id,omitempty"`            // 分派分组 ID（可选，LLM 分析指派）
	Description      string       `json:"description,omitempty"`                 // 路由描述
	WorkGroups       []RouteGroup `json:"work_groups,omitempty" gorm:"foreignKey:RouteID"`
}

// RouteGroup 路由工作分组 - 每个路由下的分类规则
type RouteGroup struct {
	ID          int    `json:"id" gorm:"primaryKey"`
	RouteID     int    `json:"route_id" gorm:"not null;index:idx_route_category,unique"`
	GroupID     int    `json:"group_id" gorm:"not null"`                       // 关联的分组 ID
	Category    string `json:"category" gorm:"not null;index:idx_route_category,unique"` // 分类名称（如"工具调用"、"编码"）
	Description string `json:"description,omitempty"`                          // 分类描述
	Keywords    string `json:"keywords,omitempty"`                             // 关键词（逗号分隔，用于匹配）
}

// RouteCreateRequest 创建路由请求
type RouteCreateRequest struct {
	Name            string                `json:"name" binding:"required"`
	PrimaryGroupID  int                   `json:"primary_group_id" binding:"required"`
	DispatchGroupID *int                  `json:"dispatch_group_id,omitempty"`
	Description     string                `json:"description,omitempty"`
	WorkGroups      []RouteGroupAddRequest `json:"work_groups,omitempty"`
}

// RouteUpdateRequest 更新路由请求 - 仅包含变更的数据
type RouteUpdateRequest struct {
	ID              int                      `json:"id" binding:"required"`
	Name            *string                  `json:"name,omitempty"`
	PrimaryGroupID  *int                     `json:"primary_group_id,omitempty"`
	DispatchGroupID *int                     `json:"dispatch_group_id,omitempty"`
	Description     *string                  `json:"description,omitempty"`
	WorkGroupsToAdd    []RouteGroupAddRequest    `json:"work_groups_to_add,omitempty"`
	WorkGroupsToUpdate []RouteGroupUpdateRequest `json:"work_groups_to_update,omitempty"`
	WorkGroupsToDelete []int                    `json:"work_groups_to_delete,omitempty"`
}

// RouteGroupAddRequest 新增工作分组请求
type RouteGroupAddRequest struct {
	GroupID     int    `json:"group_id" binding:"required"`
	Category    string `json:"category" binding:"required"`
	Description string `json:"description,omitempty"`
	Keywords    string `json:"keywords,omitempty"`
}

// RouteGroupUpdateRequest 更新工作分组请求
type RouteGroupUpdateRequest struct {
	ID          int    `json:"id" binding:"required"`
	GroupID     int    `json:"group_id,omitempty"`
	Category    string `json:"category,omitempty"`
	Description string `json:"description,omitempty"`
	Keywords    string `json:"keywords,omitempty"`
}
